pkg/kv/kvserver/replica_application_state_machine.go

```diff
@@ -223,22 +244,52 @@ func (sm *replicaStateMachine) ApplySideEffects(
 			sm.r.handleReadWriteLocalEvalResult(ctx, *cmd.localResult)
 		}
 
-		if higherReproposalsExist := cmd.proposal.Supersedes(cmd.Cmd.MaxLeaseIndex); higherReproposalsExist {
-			// If the command wasn't rejected, we just applied it and no higher
-			// reproposal must exist (since that one may also apply).
+		if higherReproposalsExist := cmd.proposal.Supersedes(cmd.Cmd.MaxLeaseIndex); higherReproposalsExist &&
+			(!cmd.Rejected() || cmd.Rejection == kvserverbase.ProposalRejectionIllegalLeaseIndex) {
+			// If this entry was accepted, there must not be superseding reproposal,
+			// as nothing prevents it from applying as well in the common case.
+			//
+			// If this entry was rejected but only on account of the lease applied
+			// index, then the call to tryReproposeWithNewLeaseIndex[^1] must have
+			// returned an error (or the proposal would not be IsLocal() now; we
+			// unbind on success). But that call, by design, does not an error for a
+			// proposal that is already superseded.
+			//
+			// If the command was rejected permanently, a superseding entry *can*
+			// actually exist, because the condition that leads to a permanent
+			// rejection now may not have been given at the time at which the
+			// superseding proposal was spawned.
 			//
-			// If the command was rejected with ProposalRejectionPermanent, no higher
-			// reproposal should exist (after all, whoever made that reproposal should
-			// also have seen a permanent rejection).
+			// For example, consider the following initial situation:
 			//
-			// If it was rejected with ProposalRejectionIllegalLeaseIndex, then the
-			// subsequent call to tryReproposeWithNewLeaseIndex[^1] must have returned an
-			// error (or the proposal would not be IsLocal() now). But that call
-			// cannot return an error for a proposal that is already superseded
-			// initially.
+			//   [lease seq is 1]
+			//   idx 99: unrelated cmd at LAI 10000, lease seq = 1
+			//   idx 100: cmd X at LAI 10000, lease seq = 1
+			//   idx 100: cmd X at LAI 10000, lease seq = 1
+			//
+			// Command `X` at log position 100 will get reproposed with a new lease
+			// index (as idx 99 consumed the LAI). But now the lease might also change
+			// hands, and an identical reproposal of index 100 might be appended, too:
+			//
+			//   [lease seq is 1]
+			//   idx 99: unrelated cmd at LAI 10000, lease seq = 1
+			//   idx 100: cmd X at LAI 10000, lease seq = 1
+			//   idx 101: new lease seq=2
+			//   idx 102: cmd X at LAI 10000, lease seq = 1
+			//   idx 103: cmd X at LAI 20000, lease seq = 1
```

Comment:
Why 2 reproposals here? Is this because there was one before and one after the lease change? So the duplicate `idx 100` above is an in-flight `idx <TBD>` reproposal?